applications/serial_lte_modem/src/slm_at_socket.c

```diff
@@ -330,13 +343,16 @@ static int do_socketopt_set(int option, int value)
 
 	/** NCS extended socket options */
 	case SO_SILENCE_ALL:
+		ret = do_nrf_setsockopt(NRF_IPPROTO_ALL, option, value);
+		break;
 	case SO_IP_ECHO_REPLY:
+		ret = do_nrf_setsockopt(NRF_IPPROTO_IP, option, value);
+		break;
 	case SO_IPV6_ECHO_REPLY:
+		ret = do_nrf_setsockopt(NRF_IPPROTO_IPV6, option, value);
+		break;
 	case SO_TCP_SRV_SESSTIMEO:
-		ret = setsockopt(sock.fd, SOL_SOCKET, option, &value, sizeof(int));
-		if (ret < 0) {
-			LOG_ERR("setsockopt(%d) error: %d", option, -errno);
-		}
+		ret = do_nrf_setsockopt(NRF_IPPROTO_TCP, option, value);
```

Comment:
This is already done by the offloading layer in `nrf/lib/nrf_modem_lib/nrf91_sockets.c` when calling `setsockopt`, we shouldn't do it here.

Reply:
Offloading layer works only for SOL_SOCKET and SOL_TLS level options.

Reply:
yea, it's a bit different. All socket options are supported by the offloading layer but for "historical" reasons the NCS socket levels are different from the native levels in `nrf_modem`. For example, `NRF_SO_IP_ECHO_REPLY` level is `NRF_IPPROTO_IP` in `nrf_modem`, but in NCS the level for `SO_IP_ECHO_REPLY` is `SOL_SOCKET`.

This is something that we could improve now, so that the documentation in `nrfxlib` can be used for NCS socket options and levels as well, and we won't need to do any translation/offloading in this module by relying on implicit knowledge of what are the expected NCS socket levels for each socket option.

@MirkoCovizzi could take a look at that during next week, but if you @MarkusLassila have some time you could update the `nrf91_socket.c` file yourself, and Mirko and myself could help you with the review.

Reply:
Currently nrf91_sockets.c:z_to_nrf_optname expects SOL_SOCKET level for the said options and ip.c:nrf_setsockopt expects the correct levels for the said options. There is no level transition which would take care of this. Has this worked at some point of time?

I think that you are suggesting that we would remove the options with incorrect levels from  nrf91_sockets.c:z_to_nrf_optname, but are you also suggesting that the missing levels (IPPROTO_TCP etc.) should be added in there? I think that this would make sense. I think that the level options are already existing in Zephyr, except the NRF_IP_PROTO_ALL (512), which is not something that exists in Linux kernel, so it most likely cannot be put in socket.h, but could be in socket_ncs.h?

Also, nrf_modem seems to support some of the options (at least PPROTO_TCP:TCP_NODELAY, IPPROTO_IP:IP_TOS) that zephyr is supporting as well, but those are not documented in [the link that you provided](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/nrf_modem/doc/sockets.html#socket-options). 
Similarly, there are other options that are supported, but not available in the documentation. This makes the situation a bit more messy than it would need to be. Should the support for those options be added as well?

I think that I could take a crack at this, it would be my first contribution to libmodem.

Reply:
@lemrey: Created a draft PR for further discussion: https://github.com/nrfconnect/sdk-nrf/pull/11715.

Reply:
We don't need changes to libmodem, only to the `nrf91_socket.c` file in NCS.